rfcs/0024-Non-consensus-node-protocol.md

```diff
@@ -0,0 +1,219 @@
+
+An NC node can delegate an account's state using the existing command `coda client delegate-stake`
+
+##### Block header at some block k (root of the transition frontier)
+
+For wallets to be able to show if transactions are finalized, non-consensus nodes will need to track the state of block at some length `k` such that the probablity of a fork of length `k` is extremely low.
+
+On startup, a non-consensus node will request from its peers the block header at root and subsequent changes in the root need to be gossiped (layer 2). Given the root and the best tip, NC nodes should be able to verify that they are indeed on the same chain. This is currently done using a merkle list of state hashes between the root and the best tip and the same can be used for non-consensus nodes as well.
+
+With the block header at the root, NC nodes can query the account state from the staged ledger materialized from the snarked ledger.
+
+The current value of k is 2160 which means for a transaction to be finalized, it would take 10 days. That does not seem like a good user experience. Transactions on Bitcoin and Ethereum are finalized after 6 (~an hour) and 30 (~7mins) block respectively.
+
+The probability that there will be a fork of length k:
+
+| k | fork probability | duration (hrs) (block_window_duration = 6mins)|
+|---|----------|----------|
+| 5 | 0.49024 | 0.5|
+| 30| 0.313376 | 3|
+| 100| 0.110262|10|
+| 200| 0.0273769 | 20 |
+| 300| 0.00697404 | 30 |
+| 400| 0.00179144 | 40 |
+| 500| 0.000461763 | 50 |
+| 600| 0.000119216 | 60 |
+| 700| 3.0803e-05 | 70 |
+| 800| 7.96218e-06 | 80 |
+| 900| 2.05858e-06 | 90 |
+| 1000| 5.32297e-07|100 |
+
+(\(\epsilon\) = 0.4)
+
+k (for non-consensus nodes) = 100?
+
+The block window duration would change with the hash function change which would reduce the wait time for finalization?
+
+### Gossip layer
+
+The gossip protocol for non-ceonsensus nodes should be separated from the one for full nodes because messages like blocks, transactions, and snark works are not required and having a different gossip layer for messages pertaining to the non-consensus nodes will reduce the bandwidth usage significantly.
+
+    1. Layer 1: Existing gossip layer that includes broadcasting blocks, transaction pool diffs, and snark pool diffs.
+    2. Layer 2: New layer for gossiping information needed by non-consensus nodes- 
+        a) Block headers
+        b) transactions (Maybe have this as a different layer in general so that archive nodes/snark workers don't have to hear any transaction-pool diffs? which means you could have one for snark-works as well to which snark workers would publish and block-producers would subscribe. This basically translates to a map of list of IPs for each topic string, overkill?)
```

Comment:
I think we’d want txns gossiped only on layer 1 (or if we added the separate layer non-consensus nodes would only connect to the layer without them). For non consensus nodes we probably only want to publish txns to the nodes connected to the other layer nets, but not read — this is similar to how light nodes behave in ethereum/bitcoin and at least empirically there is no issue with what may be perceived as an incentive misalignment

Reply:
After reading further this would mean these non consensus nodes would publish only to the txn layer but not subscribe to it

Reply:
Yes, sorry, the dot diagram shows that but I did not describe it. Will do it